core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java

```diff
@@ -643,7 +668,24 @@ private DynamicFilterContext(
             this.dynamicFilters = requireNonNull(dynamicFilters, "dynamicFilters is null");
             this.lazyDynamicFilters = requireNonNull(lazyDynamicFilters, "lazyDynamicFilters is null");
             this.replicatedDynamicFilters = requireNonNull(replicatedDynamicFilters, "replicatedDynamicFilters is null");
-            dynamicFilters.forEach(filter -> taskDynamicFilters.put(filter, new ConcurrentHashMap<>()));
+            dynamicFilters.forEach(filter -> {
+                taskDynamicFilters.put(filter, new ConcurrentHashMap<>());
+                dynamicFilterConsumers.put(filter, new ArrayList<>());
+            });
+        }
+
+        void addDynamicFilterConsumer(DynamicFilterId dynamicFilterId, Consumer<Domain> domainConsumer)
+        {
+            synchronized (dynamicFilterConsumersLock) {
+                List<Consumer<Domain>> consumers = dynamicFilterConsumers.get(dynamicFilterId);
+                if (consumers == null) {
+                    // filter has been collected concurrently
```

Comment:
more accurate comment would be:
```
// filter has already been collected
```

`// filter has been collected concurrently` referred to broadcast join where any task that produced DF "completed" filter